lgc/patch/MeshTaskShader.cpp

```diff
@@ -1272,6 +1272,13 @@ void MeshTaskShader::setMeshOutputs(Value *vertexCount, Value *primitiveCount) {
   {
     m_builder->SetInsertPoint(setMeshOutputsBlock->getTerminator());
 
+    // Check if vertex count or primitive count is zero. If so, set both to zero in order to disable vertex/primitive
+    // exporting.
+    auto product = m_builder->CreateMul(vertexCount, primitiveCount);
+    auto hasZeroCount = m_builder->CreateICmpEQ(product, m_builder->getInt32(0));
```

Comment:
Better use a bitwise OR instead of the MUL here. In SALU it doesn't matter, but in VALU the OR is faster. (And LLVM doesn't optimize that itself, I just checked.)

Actually, while we're looking at this code: have you look at the resulting ISA? There needs to be at least one `readfirstlane` for the sendmsg, but if we were to do a `readfirstlane` on the vertexCount and primitiveCount we could get a net reduction of VALU instructions. I don't think LLVM is currently smart enough to do that automatically.

Reply:
Thank you. I change the codes.

> have you look at the resulting ISA? There needs to be at least one readfirstlane for the sendmsg, but if we were to do a readfirstlane on the vertexCount and primitiveCount we could get a net reduction of VALU instructions. I don't think LLVM is currently smart enough to do that automatically.

You are right. After translating Select, LLVM generate two VGPR results for vertex count and primitive count. We have to insert readfirstlane to promote them to SGPR. This is an old issue that LLVM doesn't do well in translating Select to s_select.

Reply:
> Better use a bitwise OR instead of the MUL here. 

Just find OR doesn't fulfill the purpose. It can only detect both count are 0. What we want to check is any of the counts is 0. If MUL is costly, I decide to change it to 2EQ + OR.

Reply:
Yes, I got that logic wrong, sorry about that.

About the readfirstlane, I actually meant doing the readfirstlane *before* the cmp+or+select sequence instead of afterwards. I don't know if you've tried that and ran into issues because of LLVM's code selection?

Reply:
> I actually meant doing the readfirstlane before the cmp+or+select sequence instead of afterwards.

Ah. I see. You are right. We must make a promotion here to convert vertex/primitive count to SGPRs. LLVM doesn't know we are in a single thread block and do this work for us.